File: 0102-binary-tree-level-order-traversal/solution.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number[][]}
 */
var levelOrder = function(root) {
    return _levelOrder(root, 0, []);
};

var _levelOrder = function(root, level, ans) {
    if (!root) return ans;
    if (!ans[level]) ans[level] = [];
    
    ans[level].push(root.val);

    if (root.left) ans = _levelOrder(root.left, level+1, ans);
    if (root.right) ans = _levelOrder(root.right, level+1, ans);
    
    return ans;
}

File: 0102-binary-tree-level-order-traversal/solution.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def levelOrder(self, root: Optional[TreeNode]) -> List[List[int]]:
        # ans = []

        # def dfs(node: TreeNode, depth: int):
        #     if not node: return
        #     if len(ans) <= depth: ans.append([])
        #     ans[depth].append(node.val)
        #     dfs(node.left, depth + 1)
        #     dfs(node.right, depth + 1)

        # dfs(root, 0)

        # return ans

        ans, q = [], deque()

        if root: q.append(root)

        while q:
            v = []

            for i in range(len(q)):
                node = q.popleft()
                v.append(node.val)
                if node.left: q.append(node.left)
                if node.right: q.append(node.right)

            ans.append(v)

        return ans


